classes/local/store/s3/client.php

```diff
@@ -635,4 +643,71 @@ public function define_client_check_sdk($config) {
         }
         return $output;
     }
+
+    /**
+     * Serves range request.
+     *
+     * @param  \stored_file $file    The file to send
+     * @param  object       $ranges  Object with rangefrom, rangeto and length properties.
+     * @return false If couldn't get data.
+     * @throws \coding_exception
+     */
+    public function serve_range_request(\stored_file $file, $ranges) {
+        // Do not serve files with size less than 2MB.
+        if ($file->get_filesize() < self::MAX_TEMP_LIMIT) {
+            return false;
+        }
+
+        $response = $this->curl_range_request_to_presigned_url($file->get_contenthash(), $ranges, headers_list());
+        $httpcode = manager::get_header($response['responseheaders'], 'HTTP/1.1');
+
+        if ($response['content'] != '' && $httpcode == '206 Partial Content') {
+            header('HTTP/1.1 206 Partial Content');
+            header('Accept-Ranges: bytes');
+            $contentrange = manager::get_header($response['responseheaders'], 'Content-Range');
+            if ($contentrange !== '') {
```

Comment:
I think these 2 if statements could be simpler, it should always trust the headers from the s3 response. If those headers don't exist then don't fall back to $contentrange, just treat that the same as the other types of failures if the code is not 206. If they don't have these headers they are a totally invalid response

https://tools.ietf.org/html/rfc7233#section-4.1

> If a single part is being transferred, the server generating the 206 response MUST generate a Content-Range header field